File: leetcode3/이진희/785. Is Graph Bipartite?.java
/*

1. 아이디어 : 인접한 노드는 같은 그룹이 되면 안된다. bfs 로 모든 시작점을 탐색해서, 인접한 노드끼리 같은 그룹일 경우 false

2. 시간복잡도 : O(N+E)

3. 자료구조/알고리즘 : BFS

 */

class Solution {
    private List<List<Integer>> g = new ArrayList<>();
    private int n;
    private int[] visited;
    public boolean isBipartite(int[][] graph) {
        
        n = graph.length;
        visited = new int[n+1];

        for(int i=0; i<n; i++) g.add(new ArrayList<>());
        
        for(int i=0; i<graph.length; i++) {
            for(int j=0; j<graph[i].length; j++) {
                g.get(i).add(graph[i][j]);
            }
        }

        // 이웃한 노드끼리 다른 그룹

        for(int i=0; i<n; i++) {
            if(visited[i] != 0) continue;
            if(!bfs(i)) return false; 
        }

        return true;
    }

    private boolean bfs(int start) {
        Deque<Integer> dq = new ArrayDeque<>();

        dq.add(start);
        visited[start] = 1;

        while(!dq.isEmpty()) {
            int num = dq.poll();
            for(int node : g.get(num)) {
                if(visited[node] == 0) {
                    visited[node] = -visited[num];
                    dq.add(node);
                }
                else if (visited[num] == visited[node]) return false; 
            }
        }  

        return true;      
    }
}

File: leetcode3/최원준/785. Is Graph Bipartite?.py
#

'''
1. 아이디어 :
Bipartite는 그래프에서 노드들을 2가지 색을 칠했을때, 인접한 노드와 다른 색일때 성립.
bfs를 통해 노드들의 색갈을 0 또는 1로 칠한다.
그래프가 끊긴 문제는 모든 노드들을 순회하게끔 한다.

2. 시간복잡도 :
    O(V + E) 노드, 간선

3. 자료구조/알고리즘 :
BFS

'''
class Solution:
    def isBipartite(self, graph: List[List[int]]) -> bool:
        n = len(graph)
        colors = [-1] * n

        for i in range(n):
            if colors[i] != -1:
                continue

            colors[i] = 0
            queue = deque()
            queue.append(i)

            while queue:
                start = queue.popleft()
                start_color = colors[start]
                
                for dest in graph[start]:
                    dest_color = colors[dest]
                    if start_color == dest_color:
                        return False

                    if dest_color == -1:
                        queue.append(dest)
                        colors[dest] = (1 + start_color) % 2
            # print(colors)
        return True

                

